### PR TITLE
PT 3183 - improved migrator for old family studies

### PR DESCRIPTION
**Pull request by allasm**
- allow more patients in a family to have relatives
- migrates more complicated families
- links patients to existing pedigrees by external ID

### CODE REVIEW

**Comment by allasm** [comment]
Note: not ready to be merged yet, development version for code review (with debug code included). If approved, will be remove debug code and squash

File: components/family-studies/migrations/src/main/java/org/phenotips/studies/family/migrations/R71498PhenoTips2155DataMigration.java [hunk 1]
```diff
@@ -176,6 +259,198 @@
...
+        }
+
+        public Map<String, String> getReferredBy()
+        {
+            return this.referredBy;
+        }
+
+        public void addReferenceBy(String referredPatientId, String referredRelationship)
+        {
+            referredBy.put(referredPatientId, referredRelationship);
```
Comment (veronikaslc) [inline components/family-studies/migrations/src/main/java/org/phenotips/studies/family/migrations/R71498PhenoTips2155DataMigration.java]: this.referredBy

File: components/family-studies/migrations/src/main/java/org/phenotips/studies/family/migrations/R71498PhenoTips2155DataMigration.java [hunk 2]
```diff
@@ -710,101 +1718,11 @@ private void setAllFamilyRefs(Set<String> familyMembers, XWikiDocument familyXDo
             BaseObject familyObject = familyXDocument.getXObject(Family.CLASS_REFERENCE, true, this.context);
             familyObject.setStringListValue("members", membersRefsList);
 
-            String probandLastName = probandData.getString(LASTNAME);
-            if (!StringUtils.isEmpty(probandLastName)) {
-                familyObject.set("external_id", probandLastName, this.context);
+            if (!StringUtils.isBlank(suggestedExternalId)) {
+                familyObject.set(FAMILY_DOCUMENT_EXTERNALID_FIELD, suggestedExternalId, this.context);
```
Comment (veronikaslc) [inline components/family-studies/migrations/src/main/java/org/phenotips/studies/family/migrations/R71498PhenoTips2155DataMigration.java]: I'm not sure if we need to set the default family external ID for families that we are migrating.

Comment (allasm) [inline components/family-studies/migrations/src/main/java/org/phenotips/studies/family/migrations/R71498PhenoTips2155DataMigration.java]: @marta- what do you think about this (setting family's external IDs for migrated families)?

Comment (allasm) [inline components/family-studies/migrations/src/main/java/org/phenotips/studies/family/migrations/R71498PhenoTips2155DataMigration.java]: As discussed, it was decided that we do not want to do this. Will update PR shortly

**Comment by allasm** [comment]
The production version of the code has been committed

**Comment by allasm** [comment]
Argh, just realized I forgot to commit the checkstyle file required for build, will do as soon as I can login to my work computer

**Comment by allasm** [comment]
Added missing file

**Pull request merged**